SimulationAgent/Simulation/DeviceActor.cs

```diff
@@ -325,66 +309,54 @@ public void MoveNext()
+                    this.deviceBootstrapLogic.Stop();
+                    this.updateDeviceStateLogic.Start();
                     break;
 
                 case Status.UpdatingReportedProperties:
-                    // UpdateState Timer is not stopped as UpdatingDeviceState needs to continue to generate random telemetry for the device
                     this.ActorStatus = nextStatus;
-                    this.log.Debug("Scheduling Reported Property updates", () => new { this.deviceId });
-                    this.propertyTimer.Setup(this.updateReportedPropertiesLogic.Run, this, reportedPropertyUpdateFrequency);
-                    this.propertyTimer.Start();
-                    this.ScheduleCancellationCheckIfRequired(reportedPropertyUpdateFrequency);
+                    this.updateReportedPropertiesLogic.Start();
+                    // Note: at this point both UpdatingDeviceState
+                    //       and UpdatingReportedProperties should be running
                     break;
 
                 case Status.SendingTelemetry:
                     this.ActorStatus = nextStatus;
-                    foreach (var message in this.messages)
-                    {
-                        var telemetryTimer = this.factory.Resolve<ITimer>();
-                        this.telemetryTimers.Add(telemetryTimer);
-                        var callContext = new SendTelemetryContext
-                        {
-                            Self = this,
-                            Message = message
-                        };
-
-                        this.log.Debug("Scheduling sendTelemetryLogic", () =>
-                            new { this.deviceId, message.Interval.TotalSeconds, message.MessageSchema.Name, message.MessageTemplate });
-
-                        telemetryTimer.Setup(this.sendTelemetryLogic.Run, callContext, message.Interval);
-                        telemetryTimer.StartIn(message.Interval);
-                    }
+                    this.sendTelemetryLogic.Start();
+                    // Note: at this point
+                    //       UpdatingDeviceState, UpdatingReportedProperties and SendingTelemetry
+                    //       should be running
                     break;
                 default:
                     this.log.Error("Unknown next status",
-                        () => new { this.deviceId, this.ActorStatus, nextStatus });
+                        () => new { this.DeviceId, this.ActorStatus, nextStatus });
                     throw new ArgumentOutOfRangeException();
             }
         }
 
+        // Check if the user asks to stop the simulation
+        private void StartCancellationCheck()
+        {
+            this.cancellationCheckTimer.Setup(CancellationCheck, checkCancellationFrequency, this);
+            this.cancellationCheckTimer.Start();
+        }
+
+        private void StopCancellationCheck()
```

Comment:
I removed StopTimers because I found the previous approach quite confusing. The main source of confusion is that the actor is not exactly a state machine, multiple states can be active at the same time, some timers can be active at the same time. The initial intent behind StopTimers, was a brute force "stop everything and start only the timer for the current state" but that has become less and less the case given that the actor is 1. updating state 2. sending messages 3. updating twin.  Open to revisit the approach, but hopefully all of this complexity will go away in the next months.